File: Backend/app/services/sms_service.py
import logging
import re

import httpx

from app.config import settings

logger = logging.getLogger(__name__)

_TWILIO_URL = "https://api.twilio.com/2010-04-01/Accounts/{sid}/Messages.json"

DEFAULT_SARTHI_ASSIGNED_BODY = (
    "Dear {{passenger_name}}, your Sarthi {{sarthi_name}} will pick you up for flight "
    "{{flight_number}} on {{pickup_date}} at {{pickup_time}}.\n\n"
    "Vehicle: {{vehicle_make}} {{vehicle_name}} ({{vehicle_number}})\n"
    "Contact: {{sarthi_phone}}\n\n"
    "— Airport Transportation"
)


def render_template(body: str, variables: dict) -> str:
    """Replace {{key}} placeholders; unknown placeholders become empty string."""
    def _replace(match):
        key = match.group(1).strip()
        return str(variables.get(key, ""))
    return re.sub(r"\{\{(\w+)\}\}", _replace, body)


async def send_sms(to: str, body: str) -> bool:
    """Send an SMS via Twilio REST API. Returns True on success, False if skipped/failed."""
    sid   = settings.twilio_account_sid
    token = settings.twilio_auth_token
    from_ = settings.twilio_from_number

    if not (sid and token and from_):
        logger.warning("Twilio not configured — SMS skipped (to=%s)", to)
        return False

    # Normalise US numbers: 10-digit → +1XXXXXXXXXX
    cleaned = re.sub(r"[\s\-()]", "", to)
    if cleaned and not cleaned.startswith("+"):
        cleaned = "+1" + cleaned.lstrip("1")

    try:
        async with httpx.AsyncClient(timeout=10) as client:
            resp = await client.post(
                _TWILIO_URL.format(sid=sid),
                auth=(sid, token),
                data={"From": from_, "To": cleaned, "Body": body},
            )
            resp.raise_for_status()
        logger.info("SMS sent to %s", cleaned)
        return True
    except Exception as exc:
        logger.error("SMS failed to %s: %s", cleaned, exc)
        return False

File: Backend-js/src/services/sms_service.ts
import axios from 'axios';
import { settings } from '../config.js';
import { logger } from '../logger.js';

const TWILIO_URL = (sid: string) =>
  `https://api.twilio.com/2010-04-01/Accounts/${sid}/Messages.json`;

export const DEFAULT_SARTHI_ASSIGNED_BODY =
  'Dear {{passenger_name}}, your Sarthi {{sarthi_name}} will pick you up for flight ' +
  '{{flight_number}} on {{pickup_date}} at {{pickup_time}}.\n\n' +
  'Vehicle: {{vehicle_make}} {{vehicle_name}} ({{vehicle_number}})\n' +
  'Contact: {{sarthi_phone}}\n\n' +
  'Track live: {{tracking_url}}\n\n' +
  '- SPS Airport Transportation Team';

export const DEFAULT_INTAKE_CONFIRMATION_SMS_BODY =
  'Hi {{passenger_name}}, we got your transportation request for Suhradam Parivar Shibir. ' +
  'Track your Sarthi assignment here: {{tracking_url}}\n' +
  'Ref: {{reference}}';

export function renderTemplate(body: string, variables: Record<string, unknown>): string {
  return body.replace(/\{\{(\w+)\}\}/g, (_m, key: string) => {
    const v = variables[key.trim()];
    return v === undefined || v === null ? '' : String(v);
  });
}

function normalisePhone(to: string): string {
  let cleaned = to.replace(/[\s\-()]/g, '');
  if (cleaned && !cleaned.startsWith('+')) {
    cleaned = '+1' + cleaned.replace(/^1+/, '');
  }
  return cleaned;
}

async function twilioSend(from: string, to: string, body: string, channel: 'sms' | 'whatsapp'): Promise<boolean> {
  const sid = settings.twilio_account_sid;
  const token = settings.twilio_auth_token;
  if (!(sid && token && from)) {
    logger.warn(`Twilio not configured — ${channel} skipped (to=${to})`);
    return false;
  }
  try {
    const form = new URLSearchParams();
    form.set('From', from);
    form.set('To', to);
    form.set('Body', body);
    const resp = await axios.post(TWILIO_URL(sid), form.toString(), {
      auth: { username: sid, password: token },
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      timeout: 10_000,
      validateStatus: () => true,
    });
    if (resp.status < 200 || resp.status >= 300) {
      logger.error(`${channel} failed to ${to}: HTTP ${resp.status} — ${JSON.stringify(resp.data)}`);
      return false;
    }
    logger.info(`${channel} sent to ${to}`);
    return true;
  } catch (exc) {
    logger.error({ exc }, `${channel} failed to ${to}`);
    return false;
  }
}

export async function sendSms(to: string, body: string): Promise<boolean> {
  const from = settings.twilio_from_number;
  if (!from) {
    logger.warn(`No TWILIO_FROM_NUMBER configured — SMS skipped (to=${to})`);
    return false;
  }
  const cleaned = normalisePhone(to);
  return twilioSend(from, cleaned, body, 'sms');
}

export async function sendWhatsApp(to: string, body: string): Promise<boolean> {
  const from = settings.twilio_whatsapp_from;
  if (!from) {
    logger.info(`No TWILIO_WHATSAPP_FROM configured — WhatsApp skipped (to=${to})`);
    return false;
  }
  const cleaned = normalisePhone(to);
  const waFrom = from.startsWith('whatsapp:') ? from : `whatsapp:${from}`;
  const waTo = `whatsapp:${cleaned}`;
  return twilioSend(waFrom, waTo, body, 'whatsapp');
}

export type NotifyResult = { whatsapp: boolean };

// Passenger-facing notify: WhatsApp only. SMS is NOT attempted here by design — passengers
// communicate via WhatsApp (cheap) or email fallback (handled by caller). For admin/internal
// SMS use cases, call sendSms() directly.
export async function notify(to: string, body: string): Promise<NotifyResult> {
  const whatsapp = await sendWhatsApp(to, body);
  if (whatsapp) {
    logger.info(`notify via WhatsApp succeeded for ${to}`);
  } else {
    logger.info(`notify via WhatsApp failed for ${to} — caller should fall back to email`);
  }
  return { whatsapp };
}
